File: RNG-objc/ViewController.m
//
//  ViewController.m
//  RNG-objc
//
//  Created by Steven Xu on 2016-08-13.
//  Copyright © 2016 Steven Xu. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UIView *resultview;
@property (strong, nonatomic) IBOutlet NSLayoutConstraint *resultViewHeight;
@property (strong, nonatomic) IBOutlet NSLayoutConstraint *resultViewWidth;
@property (weak, nonatomic) IBOutlet UITextField *fromTextField;
@property (weak, nonatomic) IBOutlet UITextField *toTextField;
@property (weak, nonatomic) IBOutlet UIButton *generateBtn;

@property (readwrite) NSMutableArray<UIImage*> *imageArray;
@property CGFloat xSoFar;

@end

@implementation ViewController

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        self.imageArray = [NSMutableArray array];
        self.xSoFar = 0;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.fromTextField.delegate = self;
    self.toTextField.delegate = self;
}

- (IBAction)didTapGenerate {
    NSNumber *num = [self getRandomNum];
    [self.imageArray removeAllObjects];
    for (UIImageView *image in self.resultview.subviews) {
        [image removeFromSuperview];
    }
    self.xSoFar = 0;

    if (num == nil) {
        self.generateBtn.enabled = NO;
    } else {
        [self.view endEditing:YES];
        CGFloat rawWidth = 0;
        CGFloat rawHeight = 0;
        CGFloat maxWidth = [UIScreen mainScreen].bounds.size.width;
        CGFloat maxHeight = 120;

        for (NSUInteger i = 0; i < num.stringValue.length; i++) {
            unichar myChar = [num.stringValue characterAtIndex:i];
            UIImage *image = [UIImage imageNamed:[NSString stringWithCharacters:&myChar length:1]];
            rawWidth += image.size.width;
            rawHeight += image.size.height;
            [self.imageArray addObject:image];
        }
        rawHeight /= num.stringValue.length;

        CGFloat scale = MIN(maxWidth / rawWidth, maxHeight / rawHeight);
        CGFloat desiredHeight = rawHeight * scale;
        CGFloat desiredWidth = rawWidth * scale;

        [self.resultview layoutIfNeeded];
        [UIView animateWithDuration:1 animations:^{
            self.resultViewHeight.constant = desiredHeight;
            self.resultViewWidth.constant = desiredWidth;
            [self.resultview layoutIfNeeded];
        }];

//        NSLog(@"rawWidth: %f, rawHeight: %f, scale: %f", rawWidth, rawHeight, scale);
//        NSLog(@"desiredWidth: %f, desiredHeight: %f", desiredWidth, desiredHeight);

        for (NSUInteger i = 0; i < self.imageArray.count; i++) {
            UIImage *image = [self.imageArray objectAtIndex:i];
            CGFloat width = desiredWidth / num.stringValue.length;
            UIImageView *iv = [[UIImageView alloc] initWithFrame:CGRectMake(self.xSoFar, 0, width, desiredHeight)];
            iv.image = image;
            [self.resultview addSubview:iv];
            self.xSoFar += width;
        }
    }
}

- (NSNumber*)getRandomNum {
    NSString *from = self.fromTextField.text;
    NSString *to = self.toTextField.text;
    if (from != nil && to != nil) {
        if (from.integerValue > to.integerValue) {
            return nil;
        }
        int randNum = arc4random_uniform(to.intValue - from.intValue + 1) + from.intValue;
        NSLog(@"randNum is: %d", randNum);
        return [NSNumber numberWithInt:randNum];
    }
    return nil;
}

// MARK: UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    NSString *from = self.fromTextField.text;
    NSString *to = self.toTextField.text;
    if (from == nil || to == nil) {
        return NO;
    }
    if (self.fromTextField.isFirstResponder) {
        from = [from stringByReplacingCharactersInRange:range withString:string];
    } else if (self.toTextField.isFirstResponder) {
        to = [to stringByReplacingCharactersInRange:range withString:string];
    }
    self.generateBtn.enabled = to.integerValue >= from.integerValue;
    return YES;
}

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    [textField setText:@""];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    return YES;
}

@end

File: Whats-for-lunch/ViewController.swift
//
//  ViewController.swift
//  Whats-for-lunch
//
//  Created by Steven Xu on 2016-02-09.
//  Copyright © 2016 Steven Xu. All rights reserved.
//

import UIKit
import Foundation

class ViewController: UIViewController, UITextFieldDelegate {
	@IBOutlet weak var fromTextField: UITextField!
	@IBOutlet weak var toTextField: UITextField!
	@IBOutlet weak var generateBtn: UIButton!
	
	var resultNum = 0

	override func viewDidLoad() {
		super.viewDidLoad()
		self.fromTextField.delegate = self
		self.toTextField.delegate = self
		let tapGesture = UITapGestureRecognizer(target: self, action: #selector(ViewController.didTapView))
		self.view.addGestureRecognizer(tapGesture)
	}

	override func viewDidAppear(_ animated: Bool) {
		super.viewDidAppear(animated)
	}

	func textFieldShouldReturn(_ textField: UITextField) -> Bool {
		return true
	}

	func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
		var from = self.fromTextField.text!
		var to = self.toTextField.text!
		if self.fromTextField.isFirstResponder {
			from = (from as NSString).replacingCharacters(in: range, with: string)
		} else if self.toTextField.isFirstResponder {
			to = (to as NSString).replacingCharacters(in: range, with: string)
		}
		self.generateBtn.isEnabled = Int(to)! >= Int(from)!
		return true
	}

	func textFieldDidBeginEditing(_ textField: UITextField) {
		textField.text = ""
	}

	@IBAction func generatePressed(_ sender: UIButton) {
		if let from = UInt32(self.fromTextField.text!), let to = UInt32(self.toTextField.text!) , to >= from {
			self.generateBtn.isEnabled = true
			self.resultNum = Int(arc4random_uniform(to - from + 1)) + Int(from)
			self.view.endEditing(true)
		}
	}

	func didTapView() {
		self.view.endEditing(true)
	}

	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
		if let identifier = segue.identifier {
			if identifier == "showResult" {
				let vc = segue.destination as! ResultViewController
				vc.modalPresentationStyle = .fullScreen
				vc.modalTransitionStyle = .partialCurl
				vc.resultNum = self.resultNum
			}
		}
	}

}
